src/ray/raylet/node_manager.cc

```diff
@@ -966,18 +966,10 @@ void NodeManager::NodeRemoved(const NodeID &node_id) {
 
   // Clean up workers that were owned by processes that were on the failed
   // node.
-  for (const auto &[_, worker] : leased_workers_) {
-    const auto owner_node_id = NodeID::FromBinary(worker->GetOwnerAddress().raylet_id());
-    RAY_CHECK(!owner_node_id.IsNil());
-    if (worker->IsDetachedActor() || owner_node_id != node_id) {
-      continue;
-    }
-    // If the leased worker's owner was on the failed node, then kill the leased
-    // worker.
-    RAY_LOG(INFO).WithField(worker->WorkerId()).WithField(owner_node_id)
-        << "The leased worker is killed because the owner node died.";
-    KillWorker(worker);
-  }
+  KillWorkersOwnedByNodeID(
+      leased_workers_,
+      [this](const std::shared_ptr<WorkerInterface> &worker) { KillWorker(worker); },
+      node_id);
```

Comment:
This PR essentially replaces `worker->IsDetachedActor()` with `worker->GetAssignedTask().GetTaskSpecification().IsDetachedActor()`.

To make the changes be tested, I followed the practice in [TestHandleReportWorkerBacklog](https://github.com/ray-project/ray/pull/50280) to extract the loop on the `leased_workers_` to a static method for unit testing.

Reply:
Should we fully remove `worker.IsDetachedActor()` and/or replace its implementation with what you have here? Looks like it's likely to cause similar bugs in the future.

Reply:
Sure. `Worker::IsDetachedActor` will no longer be used after this PR. I was originally planning to remove it in a follow-up PR, but now the removal is included here.